// react-native-ui-lib/src/incubator/panView/index.js
import _isEmpty from "lodash/isEmpty";
import React from 'react';
import { PanGestureHandler } from 'react-native-gesture-handler';
import { asBaseComponent } from "../../commons/new";
import View from "../../components/view";
import { PanningDirections, PanningDirectionsEnum } from "./panningUtil";
import useHiddenLocation from "../hooks/useHiddenLocation";
import usePanGesture, { PanViewDirections, PanViewDirectionsEnum, PanViewDismissThreshold, DEFAULT_DIRECTIONS } from "./usePanGesture";
import { useAnimatedStyle } from 'react-native-reanimated';
export { PanningDirections, PanningDirectionsEnum, PanViewDirections, PanViewDirectionsEnum, PanViewDismissThreshold };

const PanView = props => {
  const {
    directions = DEFAULT_DIRECTIONS,
    dismissible,
    animateToOrigin,
    onDismiss,
    directionLock,
    threshold,
    containerStyle,
    children,
    ...others
  } = props;
  const containerRef = React.createRef();
  const {
    onLayout,
    hiddenLocation
  } = useHiddenLocation({
    containerRef
  });
  const {
    translation,
    panGestureEvent
  } = usePanGesture({
    directions,
    dismissible,
    animateToOrigin,
    onDismiss,
    directionLock,
    threshold,
    hiddenLocation
  });
  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{
        translateX: translation.x.value
      }, {
        translateY: translation.y.value
      }]
    };
  }, []);
  return <View ref={containerRef} style={containerStyle} onLayout={onLayout}>
      <PanGestureHandler onGestureEvent={_isEmpty(directions) ? undefined : panGestureEvent}>
        <View reanimated style={animatedStyle}>
          <View {...others}>{children}</View>
        </View>
      </PanGestureHandler>
    </View>;
};

PanView.displayName = 'PanView';
PanView.directions = PanViewDirectionsEnum;
export default asBaseComponent(PanView);
